core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -407,136 +416,200 @@ class KafkaApis(val requestChannel: RequestChannel,
+      val topicsPendingPartitions = new mutable.HashMap[String, OffsetCommitResponseData.OffsetCommitResponseTopic]()
+
+      val authorizedTopics = authHelper.filterByAuthorized(
+        request.context,
+        READ,
+        TOPIC,
+        offsetCommitRequest.data.topics.asScala
+      )(_.name)
+
+      def makePartitionResponse(
+        partitionIndex: Int,
+        error: Errors
+      ): OffsetCommitResponseData.OffsetCommitResponsePartition = {
+        new OffsetCommitResponseData.OffsetCommitResponsePartition()
+          .setPartitionIndex(partitionIndex)
+          .setErrorCode(error.code)
+      }
+
+      def addTopicToResponse(
+        topic: OffsetCommitRequestData.OffsetCommitRequestTopic,
+        error: Errors
+      ): Unit = {
+        val topicResponse = new OffsetCommitResponseData.OffsetCommitResponseTopic().setName(topic.name)
+        offsetCommitResponseData.topics.add(topicResponse)
+        topic.partitions.forEach { partition =>
+          topicResponse.partitions.add(makePartitionResponse(partition.partitionIndex, error))
         }
       }
-      sendResponseCallback(errorMap.toMap)
-    } else {
-      val authorizedTopicRequestInfoBldr = immutable.Map.newBuilder[TopicPartition, OffsetCommitRequestData.OffsetCommitRequestPartition]
 
-      val topics = offsetCommitRequest.data.topics.asScala
-      val authorizedTopics = authHelper.filterByAuthorized(request.context, READ, TOPIC, topics)(_.name)
-      for (topicData <- topics) {
-        for (partitionData <- topicData.partitions.asScala) {
-          val topicPartition = new TopicPartition(topicData.name, partitionData.partitionIndex)
-          if (!authorizedTopics.contains(topicData.name))
-            unauthorizedTopicErrors += (topicPartition -> Errors.TOPIC_AUTHORIZATION_FAILED)
-          else if (!metadataCache.contains(topicPartition))
-            nonExistingTopicErrors += (topicPartition -> Errors.UNKNOWN_TOPIC_OR_PARTITION)
-          else
-            authorizedTopicRequestInfoBldr += (topicPartition -> partitionData)
+      val authorizedTopicsRequest = new util.ArrayList[OffsetCommitRequestData.OffsetCommitRequestTopic]()
+      offsetCommitRequest.data.topics.forEach { topic =>
+        if (!authorizedTopics.contains(topic.name)) {
+          // If the topic is not authorized, we add the topic and all its partitions
+          // to the response with TOPIC_AUTHORIZATION_FAILED.
+          addTopicToResponse(topic, Errors.TOPIC_AUTHORIZATION_FAILED)
+        } else if (!metadataCache.contains(topic.name)) {
+          // If the topic is unknown, we add the topic and all its partitions
+          // to the response with UNKNOWN_TOPIC_OR_PARTITION.
+          addTopicToResponse(topic, Errors.UNKNOWN_TOPIC_OR_PARTITION)
+        } else {
+          // Otherwise, we check all partitions to ensure that they all exist.
+          var topicRequestWithValidPartitions: OffsetCommitRequestData.OffsetCommitRequestTopic = null
```

Comment:
Coming from reviewing OffsetFetch, I noticed we didn't do this copying into a list (or at least it didn't take up as much space). Just curious if this is because of how the old interfaces (ie, method sigatures) were different between the two operations. And if there was a way to make this less verbose.

Reply:
The big difference here is that we validate the topic and partitions exists. This validation is not present in OffsetFetch.

Reply:
Ah interesting. For offsetfetch, do we basically try to fetch and if its missing we return the error there?

Reply:
Yeah, we basically return partitions if they have committed offsets or an error.